.github/workflows/ci.yaml

```diff
@@ -7,112 +7,57 @@ on:
-          submodules: true
-
-      - name: Install Go
-        uses: actions/setup-go@v1
-        with:
-          go-version: ${{ env.GO_VERSION }}
-
-      - name: Run tests
-        shell: bash
-        run: make test
-
   build:
-    runs-on: ubuntu-20.04
-    needs: test
+    runs-on: ubuntu-22.04
     steps:
-      # submodule needs .git folder, which is missing without installing a newer git command
-      # https://github.com/actions/checkout/issues/335
-      - name: "Install latest `git`"
-        run: |
-          sudo apt purge git -y
-          sudo apt-get update && sudo apt-get install -y software-properties-common make
-          sudo add-apt-repository ppa:git-core/ppa -y
-          # apt update fails to fetch some repo due to cert failure. Skip them.
-          sudo apt update || true; sudo apt install -y --no-install-recommends git
-
       - name: Check out code
-        uses: actions/checkout@v2.3.4
-        with:
-          fetch-depth: 0
-          submodules: true
+        uses: actions/checkout@v3
 
       - name: Install Go
         uses: actions/setup-go@v3
         with:
           go-version: ${{ env.GO_VERSION }}
+          cache: true
 
-      - name: Install TinyGo
+      - name: Install TinyGo and wabt
         run: |
-          wget https://github.com/tinygo-org/tinygo/releases/download/v${TINYGO_VERSION}/tinygo_${TINYGO_VERSION}_amd64.deb
-          sudo dpkg -i tinygo_${TINYGO_VERSION}_amd64.deb
-          export PATH=$PATH:/usr/local/bin
-
-      - name: "Cache generated .wasm file"
-        uses: actions/cache@v2
+          gh release download v${TINYGO_VERSION} -p '*.linux-amd64.tar.gz' -D ~ -R github.com/tinygo-org/tinygo
+          tar -xf ~/tinygo${TINYGO_VERSION}.linux-amd64.tar.gz  -C $HOME
+          echo "$HOME/tinygo/bin" >> $GITHUB_PATH
+
+          gh release download ${WABT_VERSION} -p '*-ubuntu.tar.gz' -D ~ -R github.com/WebAssembly/wabt
+          tar -xf ~/wabt-${WABT_VERSION}-ubuntu.tar.gz  -C $HOME
+          echo "$HOME/wabt-${WABT_VERSION}/bin" >> $GITHUB_PATH
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Cache TinyGo build
+        uses: actions/cache@v3
         with:
           path: |
-            build/main.wasm
-          key: ${{ runner.os }}-cache-build-${{ github.sha }}
+            ~/.cache/tinygo
+          key: ${{ runner.os }}-tinygo-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-tinygo-
+
+      - name: Run tests
+        run: make test
 
       - name: Build WASM filter
-        shell: bash
         run: make build
 
-      - name: Install WASM
-        uses: actions-rs/toolchain@v1
-        with:
-          toolchain: stable
-
-      - name: Install Cargo
-        uses: actions-rs/cargo@v1
-        with:
-          command: install
-          args: wasm-tools
-
-      - name: Validate WASM output
-        shell: bash
-        run: wasm-tools validate build/main.wasm
-
-  e2e-test:
```

Comment:
Don't think there is any practical advantage to multiple dependent jobs, so not doing it allows us to avoid marshaling artifacts back and forth